volatility3/framework/plugins/windows/unhooked_system_calls.py

```diff
@@ -0,0 +1,186 @@
+        ]
+
+    def _gather_code_bytes(
+        self,
+        kernel: interfaces.context.ModuleInterface,
+        found_symbols: pe_symbols.found_symbols_type,
+    ) -> _code_bytes_type:
+        """
+        Enumerates the desired DLLs and function implementations in each process
+        Groups based on unique implementations of each DLLs' functions
+        The purpose is to detect when a function has different implementations (code)
+        in different processes.
+        This very effectively detects code injection.
+        """
+        code_bytes: unhooked_system_calls._code_bytes_type = {}
+
+        procs = pslist.PsList.list_processes(
+            context=self.context,
+            layer_name=kernel.layer_name,
+            symbol_table=kernel.symbol_table_name,
+        )
+
+        for proc in procs:
+            try:
+                proc_id = proc.UniqueProcessId
+                proc_name = utility.array_to_string(proc.ImageFileName)
+                proc_layer_name = proc.add_process_layer()
+            except exceptions.InvalidAddressException:
+                continue
+
+            for dll_name, functions in found_symbols.items():
+                for func_name, func_addr in functions:
+                    try:
+                        fbytes = self.context.layers[proc_layer_name].read(
+                            func_addr, 0x20
+                        )
+                    except exceptions.InvalidAddressException:
+                        continue
+
+                    if dll_name not in code_bytes:
+                        code_bytes[dll_name] = {}
+
+                    if func_name not in code_bytes[dll_name]:
+                        code_bytes[dll_name][func_name] = {}
+
+                    if fbytes not in code_bytes[dll_name][func_name]:
+                        code_bytes[dll_name][func_name][fbytes] = []
+
+                    code_bytes[dll_name][func_name][fbytes].append((proc_id, proc_name))
+
+        return code_bytes
+
+    def _generator(self) -> Generator[Tuple[int, Tuple[str, str, int]], None, None]:
+        kernel = self.context.modules[self.config["kernel"]]
+
+        found_symbols = pe_symbols.PESymbols.addresses_for_process_symbols(
+            self.context,
+            self.config_path,
+            kernel.layer_name,
+            kernel.symbol_table_name,
+            unhooked_system_calls.system_calls,
+        )
+
+        # code_bytes[dll_name][func_name][func_bytes]
+        code_bytes = self._gather_code_bytes(kernel, found_symbols)
+
+        for functions in code_bytes.values():
+            for func_name, cbb in functions.items():
+                cb = list(cbb.values())
+
+                # same implementation in all
+                if len(cb) == 1:
+                    yield 0, (func_name, "", len(cb[0]))
+                else:
+                    # find the processes that are hooked for reporting
+                    max_idx = 0 if len(cb[0]) > len(cb[1]) else 1
```

Comment:
This is impossible for someone to figure out what's going on without referring back to the previous function.  Not a show stopper, but I really, *really* dislike indexing return values, rather than using a dictionary with names, or a named tuple or similar.  There's not really a lot of point not using a lightweight tuple or dataclass when it could improve readability so much...  5:S  Again, not a showstopper, but please consider just making a dataclass to give these attribute names rather than the unfathomable `0` and `1`...  it's really not fair on readers of the code to make them go hunting...

Reply:
Ok, I used those here https://github.com/volatilityfoundation/volatility3/pull/1262 and can use them going forward.

Reply:
Ok, I think I didn't follow exactly what cb was.  It looks like we have a perfectly good dictionary, that we turn into a list of the values, and then do a comparison against that?  Do we not know the keys that the dictionary contains?  I'm, just not understanding how this code works?  So more comments please...

Reply:
check now. I pushed with many more comments